qiskit/transpiler/passes/layout/vf2_layout.py

```diff
@@ -0,0 +1,87 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""TODO A pass for choosing a Layout of a circuit onto a Coupling graph, as a
+Constraint Satisfaction Problem. It tries to find a solution that fully
+satisfy the circuit, i.e. no further swap is needed. If no solution is
+found, no ``property_set['layout']`` is set.
+"""
+import random
+from retworkx import PyGraph, PyDiGraph, graph_vf2_mapping
+from qiskit.transpiler.layout import Layout
+from qiskit.transpiler.basepasses import AnalysisPass
+
+
+class VF2Layout(AnalysisPass):
+    """If possible, chooses a Layout as a Subgraph Isomorphism Probrem, using VF2."""
+
+    def __init__(self, coupling_map, strict_direction=False, seed=None):
+        """If possible, chooses a Layout as a Subgraph Isomorphism Probrem, using VF2.
+
+        If not possible, does not set the layout property. In all the cases,
+        the property `VF2Layout_stop_reason` will be added with one of the
+        following values:
+
+        * solution found: If a perfect layout was found.
+        * nonexistent solution: If no perfect layout was found and every combination was checked.
+
+        Args:
+            coupling_map (Coupling): Directed graph representing a coupling map.
+            strict_direction (bool): If True, considers the direction of the coupling map.
+                                     Default is False.
+            seed (int): Sets the seed of the PRNG. -1 Means no node shuffling.
+        """
+        super().__init__()
+        self.coupling_map = coupling_map
+        self.strict_direction = strict_direction
+        self.seed = seed
+
+    def run(self, dag):
+        """run the layout method"""
+        qubits = dag.qubits
+        interactions = list()
+
+        for gate in dag.two_qubit_ops():
+            interactions.append((qubits.index(gate.qargs[0]), qubits.index(gate.qargs[1])))
```

Comment:
I would avoid doing it this way as using index in a loop will potentially traverse the list fully every time. Instead I would do a single iteration then just do a dict lookup to get the index with something like:

```suggestion
        qubit_indices = {qubit: index for index, qubit in enumerate(qubits)}
        interactions = [(qubit_indices[gate.qargs[0]], qubit_indices[gate.qargs[1]]) for gate in dag.two_qubit_ops()]
```

```python

```

Reply:
It makes sense, thanks! (this code is also in `CSPLayout`).